hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBindJoinSPARQLwithBoundJoin.java

```diff
@@ -0,0 +1,308 @@
+			// with values based on the incoming solution mapping
+			SPARQLGraphPattern patternWithBindings;
+			try {
+				patternWithBindings = query.applySolMapToGraphPattern( new SolutionMappingImpl(solMap) );
+			} catch ( VariableByBlankNodeSubstitutionException e ) {
+				throw new ExecOpExecutionException(e, this);
+			}
+			
+			// Create new variable 
+			final Var v = Var.alloc( renamedVar.getVarName() + "_" + i );
+			
+			// Rename the variable in the pattern
+			final Element elt2 = renameVar(patternWithBindings, renamedVar, v);
+			union.addElement(elt2);
+			
+			// Add the solution mapping to the list of solution mappings
+			solMapsList.add(solMap);
+			// Map the renamed var to the index of the corresponding solution mapping
+			renamedVars.put(v, i);
+			i++;
+		}
+		return union;
+	}
+
+	@Override
+	protected MyIntermediateResultElementSink createMySink() {
+		if (useOuterJoinSemantics) {
+			return new MyIntermediateResultElementSinkOuterJoin2(currentBatch);
+		} else {
+			return new MyIntermediateResultElementSink2(currentBatch);
+		}
+	}
+
+	// ------- helper classes ------
+
+	protected class MyIntermediateResultElementSink2 extends MyIntermediateResultElementSink
+	{
+		protected final Set<SolutionMapping> inputSolutionMappingsWithoutJoinPartners = new HashSet<>();
+
+		public MyIntermediateResultElementSink2( final Iterable<SolutionMapping> inputSolutionMappings ) {
+			super(inputSolutionMappings);
+		}
+
+		protected void _send( final SolutionMapping smFromRequest ) {
+			// Merge smFromRequest into the input solution mappings:
+			// 1. For each renamed variable, check if the request binding contains it.
+			// 2. Rename that variable and merge with the corresponding solMapList entry.
+			// 3. For each compatible input mapping, merge and collect the results.
+			for( final Entry<Var, Integer> entry : renamedVars.entrySet() ){
+				final Var v = entry.getKey();
+				final int i = entry.getValue();
+				final Binding b = smFromRequest.asJenaBinding();
+
+				if ( ! b.contains(v) ) {
+					continue;
+				}
+
+				// Rename var and merge with input mapping
+				final Binding renamedAndMerged = BindingLib.merge( renameVar(b, v, renamedVar),
+				                                                             solMapsList.get(i) );
+				final SolutionMapping updatedRequest = new SolutionMappingImpl(renamedAndMerged);
+
+				// Merge with inputSolutionMappings
+				for ( final SolutionMapping smFromInput : inputSolutionMappings ) {
+					if ( SolutionMappingUtils.compatible(smFromInput, updatedRequest ) ) {
+						solMapsForOutput.add( SolutionMappingUtils.merge(smFromInput, updatedRequest) );
+					}
+				}
+			}
+		}
+	} // end of helper class MyIntermediateResultElementSink2
+
+
+	protected class MyIntermediateResultElementSinkOuterJoin2 extends MyIntermediateResultElementSink2
+	{
+		protected final Set<SolutionMapping> inputSolutionMappingsWithJoinPartners = new HashSet<>();
```

Comment:
A better option is to have `Set<Integer>` here. Then, within the `_send` method, after extracting the integer from the renamed variable in the given `smFromRequest`, you add that integer to this set. Thereafter, in `flush()`, you iterate over all integers from 0 to `solMapsList.size()-1`; each of these integers that is not in this set is the index of a solution mapping in `solMapsList` that did not have a join partner and, thus, needs to be added to `solMapsForOutput`.

Reply:
As in the case of `_send` above, `solMapsList` contains only the restricted solution mappings so we still need to loop over `inputSolutionMappings`.